### PR TITLE
Fix #22080: Detect anacrusis for beaming and grouping

### LINKED ISSUE #22080: incorrect beaming in pickup measure
### Issue type

Engraving bug (incorrect score rendering)

### Bug description

If you have a score with a pickup maesure on an unaccented beat the auto beaming is wrong.

### Steps to reproduce

1. Create a new 6/8 score with a pickup measure 4/8
2. Write 4 eight Notes
Wrong:
image

Right:
image

### Screenshots/Screen recordings

_No response_

### MuseScore Version

OS: Windows 10 Version 2009 or later, Arch.: x86_64, MuseScore version (64-bit): 4.2.1-240230937, revision: github-musescore-musescore-d757433

### Regression

I don't know

### Operating system

Windows 11

### Additional context

This also happens in a 3/4 measure with an 3/8 Pickup measure.

### PR DESCRIPTION
**Pull request by XiaoMigros**
Resolves: #22080

Besides beams, this PR changes how rhythms are grouped in pickup measures, meaning that rest grouping and 'regroup rhythms' are correct now as well.

- [x] I signed the [CLA](https://musescore.org/en/cla)
- [x] The title of the PR describes the problem it addresses
- [x] Each commit's message describes its purpose and effects, and references the issue it resolves
- [x] If changes are extensive, there is a sequence of easily reviewable commits
- [x] The code in the PR follows [the coding rules](https://github.com/musescore/MuseScore/wiki/CodeGuidelines)
- [x] There are no unnecessary changes
- [x] The code compiles and runs on my machine, preferably after each commit individually
- [x] I created a unit test or vtest to verify the changes I made (if applicable)

### CODE REVIEW

**Comment by MarcSabatella** [comment]
Nice!  FWIW I’d normally be concerned about effect on older scores with a change like this, but it’s not too likely this will do anything but improve things in scores that hadn’t been manually adjusted, and have no effect on scores that were.

However, it would be nice to enhance the algorithm to handle the case of a split measure that is *not* interrupted by a break or frame.  I mentioned the case of a mid-measure repeat, and the code *could* just check for that explicitly.  But it would be nice to handle this even when there isn’t a repeat.  I always imagined we could check the duration of the previous measure, and if it *also* is shorter than the time signature, assume the current measure is a continuation of it.  No doubt that’s wrong in some obscure corner case but it’s going to be right most of the time.  To be safer, maybe we could check if the two partial durations add up to the time signature, which would pretty much clinch it.

**Comment by XiaoMigros** [comment]
Maybe do a much more simple check of 'the measure is shorter than the time signature' and 'the measure is excluded from the measure count'?

**Comment by MarcSabatella** [comment]
Oooh, I like that idea!  I guess it’s possible someone might exclude the first partial measure instead of the second, but they’d be wrong :-).  I’d still keep the existing checks too because, realistically, not everyone does the exclude thing at all.

**Comment by XiaoMigros** [comment]
Has been taken care of, thanks!

**Review by cbjeukendrup — APPROVED** [review]

**Review by oktophonie — APPROVED** [review]

File: src/engraving/dom/measure.cpp [hunk 1]
```diff
@@ -3132,6 +3133,29 @@ Fraction Measure::computeTicks()
...
+{
+    Fraction offset = Fraction(0, 1);
+    const MeasureBase* pm = prev();
+    ElementType pt = pm ? pm->type() : ElementType::INVALID;
+
+    if (irregular() || !pm
+        || pm->lineBreak() || pm->pageBreak() || pm->sectionBreak()
+        || pt == ElementType::VBOX || pt == ElementType::HBOX
+        || pt == ElementType::FBOX || pt == ElementType::TBOX) {
+        if (timesig() - ticks() > Fraction(0, 1)) {
```
Comment (cbjeukendrup) [inline src/engraving/dom/measure.cpp:3152]: Shouldn't `Measure::isAnacrusis` be updated to use this same condition?

Comment (XiaoMigros) [reply src/engraving/dom/measure.cpp:3152]: > Shouldn't `Measure::isAnacrusis` be updated to use this same condition?

I was afraid that might have unwanted consequences, but I'll give that a try

> Also, do you know what's the relation/difference between `Measure::timesig()` and `score()->sigmap()->timesig(tick().ticks()).nominal()`?

As I recall, `Measure::timesig()` returns a `Fraction` value and the second method returns a `TimeSigFrac` value (which contains more detailed information needed to compute the groupings.

**Review by cbjeukendrup — APPROVED** [review]

**Comment by Jojo-Schmitz** [comment]
See #22370, on this PR too no unit tests were run

**Pull request merged**